modules/renter/download.go

```diff
@@ -1,805 +1,505 @@
 package renter
 
-// NOTE: All chunk recovery (which involves high computation and disk syncing)
-// is done in the primary download loop thread. At some point this may be a
-// significant performance bottleneck.
+// The download code follows a hopefully clean/intuitive flow for getting super
+// high and computationally efficient parallelism on downloads. When a download
+// is requested, it gets split into its respective chunks (which are downloaded
+// individually) and then put into the download heap. The primary purpose of the
+// download heap is to keep downloads on standby until there is enough memory
+// available to send the downloads off to the workers. The heap is sorted first
+// by priority, but then a few other criteria as well.
+//
+// Some downloads, in particualr downloads issued by the repair code, have
+// already had their memory allocated. These downloads get to skip the heap and
+// go straight for the workers.
```

Comment:
no, if they have to wait in the heap they can cause a deadlock, because their corresponding upload memory cannot be released, and they may get stuck waiting for another download to acquire memory.

This is effectively a priority lane. The purpose of the heap is exclusively to ratelimit the act of giving downloads to workers. By giving the pieces directly to the workers instead of putting them through the heap, this is functionally equivalent to putting them at the top of the heap and triggering the heap immediately.